### PR TITLE
patch: modify multipart logic to support attachment IDs / indexes

### PR DESCRIPTION
**Pull request by sudojunior**
* remains to be fully tested, and documented
* initial tests reveal attachments may need to retain original attachment IDs from the sent message
* specifying `attachments` as part of the message body, removes the ability to auto resolve attachment indexes

- [x] Test suite (`.send`, ~~`.sendFollowUp`~~)
- [ ] Active deployment test

<details>
<summary>Possible command for deployment test</summary>

```ts
import { CommandContext, SlashCommand, SlashCreator } from 'slash-create';

export default class FileTestCommand extends SlashCommand {
  constructor(creator: SlashCreator) {
    super(creator, {
      name: 'file-test',
      description: 'Test file upload'
    });
  }

  async run(ctx: CommandContext) {
    await ctx.defer();
    await ctx.send({
      content: "Here's a file!",
      file: {
        name: 'test.txt',
        file: Buffer.from('This is a test file.')
      }
    });

    await ctx.editOriginal({
      content: "Here's another file!",
      attachments: [
        {
          id: '0',
          name: 'test.txt'
        },
        {
          id: '1',
          name: 'test2.txt'
        }
      ],
      file: {
        name: 'test2.txt',
        file: Buffer.from('This is another test file.')
      }
    });
  }
}
```
</details>

> `Invalid form data` error originates from the edit attempt not having an ID attached to the file - due to how it utilizes attachments when building multipart data.

### CODE REVIEW

File: src/util/requestHandler.ts [hunk 1]
```diff
@@ -102,18 +102,18 @@ export class RequestHandler {
             if (Array.isArray(file)) {
               data = new MultipartData();
               headers['Content-Type'] = 'multipart/form-data; boundary=' + data.boundary;
-              file.forEach((f) => {
+              file.forEach((f, index) => {
                 if (!f.file) {
                   return;
                 }
-                (data as MultipartData).attach(f.name, f.file, f.name);
+                (data as MultipartData).attach(`files[${f.id ?? index}]`, f.file, f.name);
```
Comment (sudojunior) [inline src/util/requestHandler.ts:109]: The association between `MessageFile` and `MessageAttachmentOptions` differs for the intention of the request itself. Indicating the intention to keep a file for the PATCH request currently has to be done through `file: % | %[]` - but currently does not have anyway to handle existing files or removal of files without associating the file buffer with the mulitpart request.

> tl;dr - As the code currently is, the request has to have the file buffer to perform the request (regardless of the intention to keep the files from the message state before, or not).

Comment (sudojunior) [reply src/util/requestHandler.ts:109]: By all accounts, L106 above should be designed to handle that - but the interface forces the use of a file buffer (even as a file array).

**Comment by Snazzah** [comment]
I've updated the way attachments here are managed. It's a lot more hands off, since you can still keep existing attachments if you provide their ID. Here's an example of keeping the previous attachment and giving a description in a new file. Although updating the description of an existing file does not work.
```js
await ctx.defer();
const message = await ctx.sendFollowUp('hi', { file: { name: 'a.txt', file: Buffer.from('b') } });
const image = await axios('https://snazzah.com/images/meta/favicon-32x32.png', { responseType: 'arraybuffer' });
await ctx.editOriginal({
  attachments: [
    {
      id: message.attachments[0].id,
      description: 'this wont be updated'
    },
    {
      id: 0,
      description: 'hi'
    }
  ],
  file: { name: 'b.png', file: Buffer.from(image.data) }
});
```

Not providing the first attachment will remove that one from the message, so whatever array of attachments is sent to Discord will show up as given. (The order of new attachments doesn't matter though, attachments are sorted by their creation)
